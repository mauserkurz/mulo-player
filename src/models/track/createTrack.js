const DEFAULT_BLOB = null;

class Track {
  constructor({
    id = '',
    name = '',
    blob = DEFAULT_BLOB,
    lastModifiedDate = new Date().getTime(),
    isLoading = false,
  }) {
    this.id = id;
    this.name = name;
    this.blob = blob;
    this.lastModifiedDate = lastModifiedDate;
    this.isLoading = isLoading;
  }

  get isLoaded() {
    return this.blob !== DEFAULT_BLOB;
  }
}

export default options => new Track(options);
