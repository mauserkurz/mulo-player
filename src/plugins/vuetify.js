import Vue from 'vue';
import Vuetify from 'vuetify';
import { THEME as theme } from '@/constants';

Vue.use(Vuetify);

export default new Vuetify({ theme });
