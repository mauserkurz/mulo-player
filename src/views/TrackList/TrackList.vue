<template>
  <v-list
    class="track-list"
    rounded>
    <v-list-item-group
      color="primary">
      <v-text-field
        v-model="search"
        prepend-inner-icon="search"
        flat
        solo-inverted
        clearable
        label="Search"
        @click:clear="clearSearch"/>

      <v-list-item
        v-for="track of filteredTrackList"
        :key="track.id"
        :class="{ 'v-list-item--active': isTrackSelected(track.id) }"
        @click="switchTrack(track.id)">
        <v-list-item-icon>
          <v-icon
            v-if="!track.isLoading"
            v-text="getIcon(track.isLoaded)"
            @click.stop="getTrack(track.id)"/>
          <v-progress-circular
            v-else
            :size="24"
            indeterminate
            @click.stop="cancelGettingTrack(track.id)"/>
        </v-list-item-icon>

        <v-list-item-content>
          <v-list-item-title v-text="track.name"/>
        </v-list-item-content>
      </v-list-item>
    </v-list-item-group>
  </v-list>
</template>

<script>
export default {
  name: 'TrackList',

  props: {
    trackList: {
      type: Array,
      required: true,
    },

    currentTrack: {
      type: Object,
      required: true,
    },
  },

  data() {
    return { search: '' };
  },

  computed: {
    filteredTrackList() {
      return this.trackList.filter(({ name }) => !this.search || name.includes(this.search));
    },
  },

  methods: {
    isTrackSelected(trackID) {
      return this.currentTrack.id === trackID;
    },

    getIcon(isLoaded) {
      let icon = 'mdi-cloud-download';

      if (isLoaded) {
        icon = 'mdi-check-circle';
      }
      return icon;
    },

    switchTrack(trackID) {
      this.$emit('switch-track', trackID);
    },

    getTrack(trackID) {
      this.$emit('get-track', trackID);
    },

    cancelGettingTrack(trackID) {
      this.$emit('cancel-getting-track', trackID);
    },

    clearSearch() {
      this.search = '';
    },
  },
};
</script>

<style scoped lang="less">
.track-list {}
</style>
