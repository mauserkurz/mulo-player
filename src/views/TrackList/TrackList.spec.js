import { mount, shallowMount, createLocalVue } from '@vue/test-utils';
import createTrack from '@/models/track/createTrack';
import vuetify from '@/plugins/vuetify';
import TrackList from './TrackList.vue';

const localVue = createLocalVue();

describe('Component TrackList', () => {
  const trackList = [
    createTrack({
      id: 0,
      name: 'Band 0 - Song 0',
      isLoading: true,
    }),
    createTrack({
      id: 1,
      name: 'Band 1 - Song 1',
      blob: new Blob([], { type: 'audio/mpeg' }),
    }),
    createTrack({
      id: 2,
      name: 'Band 2 - Song 2',
    }),
  ];
  const defaultProps = {
    trackList,
    currentTrack: trackList[0],
  };
  const createWrapper = ({ isWithoutStubs = false, propsData = {}, ...options } = {}) => {
    const renderer = isWithoutStubs ? mount : shallowMount;
    const wrapper = renderer(TrackList, {
      localVue,
      vuetify,
      propsData: Object.assign({}, defaultProps, propsData),
      ...options,
    });

    return { wrapper };
  };

  describe('snapshots', () => {
    it('should match snapshot', () => {
      const { wrapper } = createWrapper({ isWithoutStubs: true });

      expect(wrapper.html()).toMatchSnapshot();
    });
  });

  describe('events', () => {
    it('should emit switch-track event after click on list item', () => {
      const { wrapper } = createWrapper({ isWithoutStubs: true });

      wrapper.findAll('.v-list-item').at(1).trigger('click');
      expect(wrapper.emitted()['switch-track']).toEqual([[1]]);
    });

    it('should emit get-track event after click on download icon', () => {
      const { wrapper } = createWrapper({
        stubs: {
          'v-icon': {
            template: '<div class="v-icon" @click="$listeners.click" />',
          },
        },
      });

      wrapper.findAll('.v-icon').at(0).trigger('click');
      expect(wrapper.emitted()['get-track']).toEqual([[1]]);
    });

    it('should emit cancel-getting-track event after click on spinner icon', () => {
      const { wrapper } = createWrapper({ isWithoutStubs: true });

      wrapper.find('.v-progress-circular').trigger('click');
      expect(wrapper.emitted()['cancel-getting-track']).toEqual([[0]]);
    });

    it('should show filtered list after search input', async () => {
      const { wrapper } = createWrapper({ isWithoutStubs: true });
      const input = wrapper.find('input[type="text"]');

      input.element.value = 'Band 1';
      input.trigger('input');
      await wrapper.vm.$nextTick();

      expect(wrapper.findAll('.v-list-item__title').length).toBe(1);
      expect(wrapper.find('.v-list-item__title').text()).toBe(trackList[1].name);
    });

    it('should clear search input after click on clear icon', async () => {
      const { wrapper } = createWrapper({ isWithoutStubs: true });
      const input = wrapper.find('input[type="text"]');

      input.element.value = 'Band 1';
      input.trigger('input');
      await wrapper.vm.$nextTick();
      wrapper.find('.v-input').vm.$emit('click:clear');
      await wrapper.vm.$nextTick();

      expect(input.element.value).toBe('');
    });
  });
});
