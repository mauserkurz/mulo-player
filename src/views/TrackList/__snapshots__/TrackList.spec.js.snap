// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component TrackList snapshots should match snapshot 1`] = `
<div role="list" class="v-list track-list v-sheet v-sheet--tile theme--light v-list--rounded">
  <div role="listbox" class="v-item-group theme--light v-list-item-group primary--text">
    <div class="v-input theme--light v-text-field v-text-field--single-line v-text-field--solo v-text-field--solo-inverted v-text-field--solo-flat v-text-field--enclosed">
      <div class="v-input__control">
        <div class="v-input__slot">
          <div class="v-input__prepend-inner">
            <div class="v-input__icon v-input__icon--prepend-inner"><i aria-hidden="true" class="v-icon notranslate material-icons theme--light">search</i></div>
          </div>
          <div class="v-text-field__slot"><label for="input-5" class="v-label theme--light" style="left: 0px; position: absolute;">Search</label><input id="input-5" type="text"></div>
          <div class="v-input__append-inner">
            <div class="v-input__icon v-input__icon--"><i role="button" tabindex="0" class="v-icon notranslate v-icon--link material-icons theme--light"></i></div>
          </div>
        </div>
        <div class="v-text-field__details">
          <div class="v-messages theme--light"><span name="message-transition" tag="div" class="v-messages__wrapper"></span></div>
        </div>
      </div>
    </div>
    <div tabindex="0" role="listitem" aria-selected="false" class="v-list-item v-list-item--link theme--light v-list-item--active">
      <div class="v-list-item__icon">
        <div role="progressbar" aria-valuemin="0" aria-valuemax="100" class="v-progress-circular v-progress-circular--indeterminate" style="height: 24px; width: 24px;"><svg xmlns="http://www.w3.org/2000/svg" viewBox="24 24 48 48" style="transform: rotate(0deg);">
            <circle fill="transparent" cx="48" cy="48" r="20" stroke-width="8" stroke-dasharray="125.664" stroke-dashoffset="125.66370614359172px" class="v-progress-circular__overlay"></circle>
          </svg>
          <div class="v-progress-circular__info"></div>
        </div>
      </div>
      <div class="v-list-item__content">
        <div class="v-list-item__title">Band 0 - Song 0</div>
      </div>
    </div>
    <div tabindex="0" role="listitem" aria-selected="false" class="v-list-item v-list-item--link theme--light">
      <div class="v-list-item__icon"><i role="button" tabindex="0" class="v-icon notranslate v-icon--link mdi mdi-check-circle theme--light"></i></div>
      <div class="v-list-item__content">
        <div class="v-list-item__title">Band 1 - Song 1</div>
      </div>
    </div>
    <div tabindex="0" role="listitem" aria-selected="false" class="v-list-item v-list-item--link theme--light">
      <div class="v-list-item__icon"><i role="button" tabindex="0" class="v-icon notranslate v-icon--link mdi mdi-cloud-download theme--light"></i></div>
      <div class="v-list-item__content">
        <div class="v-list-item__title">Band 2 - Song 2</div>
      </div>
    </div>
  </div>
</div>
`;
